### PR TITLE
Add a new match compiler implementation

### LINKED ISSUE #183: Pattern-matching a value causes Infinite loops
```
use std

type ty = union
        `Tyarray int
        `Other
;;

type srcloc = struct
        file    : byte[:]
        line    : int
;;

type tydef = struct
        ty      : ty
        loc     : srcloc
        fixed   : bool
;;

const istyarray = {t
        match t
        | [.ty=`Tyarray 1 ]:    -> true
        | _:                    -> false
        ;;
}

const main = {
        var b : tydef = [.ty = `Other]
        var v = istyarray(b)

        std.put("{}\n", v)
}

```
Tested with the master (9664143)
I expected that it should show 'false'. It's not obvious how this could generate an infinite loop.
Not an urgent issue.

### PR DESCRIPTION
**Pull request by typeless**
Fixes #183

References:
"When Do Match-Compilation Heuristics Matter?" by Kevin Scott and Norman Ramsey

Stats:
Sample count: 506
Dtree Size      avg: 5.38       95th percentile: 3.00    maximum: 100
Dtree Height    avg: 1.39       95th percentile: 1.00    maximum: 12

Sample generation:
$ MATCH_STATS=1 make bootstrap && mbld -R support/matchstats.myr ./match.csv

### CODE REVIEW

**Comment by typeless** [comment]
@oridb no problem. Just take your time. 😉

**Comment by typeless** [comment]
TODO:
1. rename _XXX variables/functions/macros in match_test.c
2. figure out how to implement the exhaustiveness checking.

**Comment by oridb** [comment]
It looks like the existing verifymatch() code will work for exhaustiveness checking as long as you modify the code to record up the number of constructors the type of a decision tree node can have in `dt->nconstructors` . The algorithm we have is dead simple -- a match is exhaustive if all constructors have a decision edge *or* there's a wild card.

We can just as easily record the type, and then call nconstructors() on that type.

**Comment by typeless** [comment]
@oridb
It does not seem to be easy as I thought because the code separates the pattern matrix construction phase from the dtree construction phase. It is a bit different than the original algorithm in the paper which produces the dtree while traversing the pattern tree. I decided to have two phases since it does not seem to be trivial to model our patterns (in particular tuples and friends) after the ‘c(p1,p2,...pn)’ form cleanly and it would simplify the following steps.  

I am looking at “ A Term Pattern-Match Compiler Inspired by Finite Automata Theory”. It has some hints about the problem. I am going to experiment the approach. (And it reminds me that I have to expand the wildcard patterns. That is another todo)

I’ll look into it tomorrow (Taiwan time).

**Comment by oridb** [comment]
There are two different errors that the current pattern match compiler provides. I think the first -- hide checking (don't know if there's a better term) will not transfer, but the second (exhaustiveness checking) will.

The hide checking currently works by simply checking as the tree is built whether adding a pattern to the decision tree actually changes the state of the decision tree under construction. If there's no change made to the tree when a new pattern is wired in, then clearly the pattern was already covered by the previous matches added in. This part will need to be rethought.

The exhaustiveness check, though, is run on  a fully constructed dtree, and does not need any cooperation from the intermediate steps. I think we can just use it as is.

**Comment by typeless** [comment]
@oridb 
You make a good point that `verifymatch` can work with the new match compiler. I gave it a second try and it is actually feasible.
The key is that we can retrieve the type information of the patterns from the `load` field indirectly, which I thought it can only be added when building the intermediate data structure.
The match compiler has `veritymatch` now.

**Comment by typeless** [comment]
What is the recommended style for naming C macros in the codebase? like https://github.com/oridb/mc/blob/3704a945087ebf308f457a56753f30e0c07a4b3b/mi/match_test.c#L398 
I am thinking about `INIT_T`, `INIT_P0` and so forth.

**Comment by oridb** [comment]
`T`, `P0`, etc are fine. I'm not really namespacing anything here. I use a slightly unconventional style, so if you wanted init in there, it'd be `InitT`, `InitP0`, 

If a longer name makes things significantly more readable, that's fine too.

**Comment by oridb** [comment]
(oops, didn't mean to close this, wrong button)

**Comment by typeless** [comment]
The hide-checking has been implemented.

File: mi/match.c [hunk 1]
```diff
@@ -163,564 +180,580 @@ mkdtree(Srcloc loc, Node *lbl)
...
-	}
-	lcat(end, nend, last, nlast);
-	lfree(&last, &nlast);
-	return ret;
+	return out;
 }
 
-static int
-addlit(Node *pat, Node *val, Dtree *start, Dtree *accept, Node ***cap, size_t *ncap, Dtree ***end, size_t *nend)
+/* compile implements the algorithm outlined in the paper "When Do Match-Compilation Heuristics Matter?" by Kevin Scott and Norman Ramsey
```
Comment (oridb) [inline mi/match.c]: Some of these lines are a bit long. I don't go for hard character limits (it's better to go over by a few chars than to wrap things oddly), but if you could wrap this comment, that would be appreciated.

File: mi/match.c [hunk 2]
```diff
@@ -163,564 +180,580 @@ mkdtree(Srcloc loc, Node *lbl)
...
+
+static size_t
+nconstructors(Type *t)
+{
+	if (!t)
+		return 0;
+
+	t = tybase(t);
+	switch (t->type) {
+	case Tyvoid:return 1;break;
```
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match.c]: Arent' these breaks redundant? I'll remove them unless it's a style requirement.

File: mi/match.c [hunk 3]
```diff
@@ -13,38 +14,72 @@
...
-	/* values for matching */
-	Node *lbl;
+static int ndtree;
+
+/* Path is a integer sequence that labels a subtree of a subject tree.
+ * For example,
+ * 0 is the root
+ * 0,0 and 0,1 are two subtrees of the root.
+ *
+ * Note: the order of the sequence is reversed with regard to the reference paper.
```
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match.c]: Done

File: mi/match.c [hunk 4]
```diff
@@ -13,38 +14,72 @@
...
+ * | (21, 22, 23)
+ * | (31, 32, 33)
+ * | _
+ *
+ * the entries 11, 21, 31 and the wildcard pattern at the bottom have the same Path 0,1
+ *             12, 22, 32 have the same Path 0,2
+ *             13, 23, 33 have the same Path 0,3
+ */
+typedef struct Path {
+	unsigned char len;
```
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match.c]: Done

File: mi/match.c [hunk 5]
```diff
@@ -13,38 +14,72 @@
...
+} Slot;
 
-	/* the decision tree step */
-	Node **pat;
-	size_t npat;
-	Dtree **next;
-	size_t nnext;
-	Dtree *any;
+static Slot *
+newslot(Path *path, Node *pat, Node *val)
```
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match.c]: Done

File: mi/match.c [hunk 6]
```diff
@@ -163,564 +180,580 @@ mkdtree(Srcloc loc, Node *lbl)
...
 
-	ret = 0;
-	if (t->nnext == t->nconstructors || t->any)
-		ret = 1;
-	for (i = 0; i < t->nnext; i++)
-		if (!verifymatch(t->next[i]))
-			ret = 0;
-	return ret;
+static Path*
+newpath(Path *p, char c)
```
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match.c]: Done

File: mi/match.c [hunk 7]
```diff
@@ -163,564 +180,580 @@ mkdtree(Srcloc loc, Node *lbl)
...
+		return liteq(a->expr.args[0], b->expr.args[0]);
+	case Ogap:
+	case Ovar:
 		return 1;
-	if (ty->type == Tystruct)
-		return ty->nmemb == 0;
-	if (ty->type == Tyarray)
-		return fold(ty->asize, 1)->expr.args[0]->lit.intval == 0;
+	default:
+		assert(0);
```
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match.c]: Done

File: mi/match.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match.c]: Done.

File: mi/match.c (hunk 1 above)
Comment (typeless) [reply mi/match.c]: Done

File: mi/match_test.c [hunk 8]
```diff
@@ -298,13 +298,13 @@ main(int argc, char **argv)
 
 #define P(x) ({D();p##x;})
 
-#define __P0(p) ({D();p0 = (p);})
-#define __P1(p) ({D();p1 = (p);})
-#define __P2(p) ({D();p2 = (p);})
-#define __P_(p) ({D();p_ = (p);})
+#define InitP0(p) ({D();p0 = (p);})
```
[the remark this answers is not in the document]
Comment (typeless) [reply mi/match_test.c]: Done

**Comment by typeless** [comment]
By the way, I don't intend to keep the commits after the PR is merged. So, feel free to squash them when merging.

**Comment by typeless** [comment]
I am working on implementing some heuristics, which I plan to submit in another PR.

**Comment by typeless** [comment]
I squash all commits into one. That makes the history cleaner. 
You would need to rewrite the history though.

**Comment by typeless** [comment]
Fixed a bug in the parsing code of the stats file.

**Pull request merged**